src/dvc_data/index/index.py

```diff
@@ -442,6 +516,106 @@ def remote_exists(self, entry: "DataIndexEntry", **kwargs) -> bool:
 
         return storage.remote.exists(entry, **kwargs)
 
+    def _bulk_storage_exists(
+        self,
+        entries: list[DataIndexEntry],
+        storage_selector: Callable[["StorageInfo"], Optional["Storage"]],
+        callback: Callback = DEFAULT_CALLBACK,
+        **kwargs,
+    ) -> dict[DataIndexEntry, bool]:
+        by_storage: dict[Optional[Storage], list[DataIndexEntry]] = defaultdict(list)
+        for entry in entries:
+            storage_info = self[entry.key]
+            storage = storage_selector(storage_info) if storage_info else None
+            by_storage[storage].append(entry)
+
+        results = {}
+
+        # Unify batches per actual underlying ODB path.
+        # Maps from (storage_type, odb_path) to [(StorageInstance, entries)]
+        odb_batches: dict[
+            tuple[type, str | None], list[tuple[ObjectStorage, list[DataIndexEntry]]]
+        ] = defaultdict(list)
+
+        for storage, storage_entries in by_storage.items():
+            if storage is None:
+                for entry in storage_entries:
+                    raise StorageKeyError(entry.key)
+                continue
+
+            if not isinstance(storage, ObjectStorage):
+                # We won't optimize this and run it normally.
+                if hasattr(storage, "bulk_exists"):
+                    storage_results = storage.bulk_exists(
+                        storage_entries, callback=callback, **kwargs
+                    )
+                    results.update(storage_results)
+                else:
+                    for entry in callback.wrap(storage_entries):
+                        results[entry] = storage.exists(entry, **kwargs)
```

Comment:
Please create `bulk_exists` with this naive implementation on the base class. We can optimize this in the future, and will also cleanup the code.

Reply:
I've added a naive `bulk_exists` to the base class.